### ISSUE TITLE
gcc4.9: can't compile grids.cpp 

### ISSUE THREAD

**Issue by SpaceIm** [OSGeo/PROJ#2095]
#### Problem description

gcc 4.9 fails to compile grids.cpp.

```
/home/conan/.conan/data/proj/7.0.0/bincrafters/testing/build/860a44e8cc899028241eb931b876e72da5bc34f2/source_subfolder/src/grids.cpp: In static member function 'static std::unique_ptr<osgeo::proj::VerticalShiftGridSet> osgeo::proj::VerticalShiftGridSet::open(PJ_CONTEXT*, const string&)':
/home/conan/.conan/data/proj/7.0.0/bincrafters/testing/build/860a44e8cc899028241eb931b876e72da5bc34f2/source_subfolder/src/grids.cpp:1379:16: error: cannot bind 'std::unique_ptr<osgeo::proj::GTiffVGridShiftSet>' lvalue to 'std::unique_ptr<osgeo::proj::GTiffVGridShiftSet>&&'
         return set;
                ^
In file included from /usr/include/c++/4.9/memory:81:0,
                 from /home/conan/.conan/data/proj/7.0.0/bincrafters/testing/build/860a44e8cc899028241eb931b876e72da5bc34f2/source_subfolder/src/grids.hpp:31,
                 from /home/conan/.conan/data/proj/7.0.0/bincrafters/testing/build/860a44e8cc899028241eb931b876e72da5bc34f2/source_subfolder/src/grids.cpp:34:
/usr/include/c++/4.9/bits/unique_ptr.h:220:2: note: initializing argument 1 of 'std::unique_ptr<_Tp, _Dp>::unique_ptr(std::unique_ptr<_Up, _Ep>&&) [with _Up = osgeo::proj::GTiffVGridShiftSet; _Ep = std::default_delete<osgeo::proj::GTiffVGridShiftSet>; <template-parameter-2-3> = void; _Tp = osgeo::proj::VerticalShiftGridSet; _Dp = std::default_delete<osgeo::proj::VerticalShiftGridSet>]'
  unique_ptr(unique_ptr<_Up, _Ep>&& __u) noexcept
  ^
```

gcc 4.9 cannot return a unique_ptr of a derived type and relies on copy ellision.

Declaration of `VerticalShiftGridSet::open` is:
```c++
std::unique_ptr<VerticalShiftGridSet> VerticalShiftGridSet::open(PJ_CONTEXT *, const std::string &)
```

If TIFF_ENABLED, it may return a `std::unique_ptr<GTiffVGridShiftSet>` through `GTiffVGridShiftSet::open`, which is usually fine because `GTiffVGridShiftSet` is derived from `VerticalShiftGridSet`. But gcc < 5 cannot handle this case.

Didn't tested but it might be fixed by using `std::move` when returning `set`.

(the problem is the same in HorizontalShiftGridSet::open and GTiffGenericGridShiftSet::open)

#### Environment Information
 - PROJ version (`proj`): 7.0.0
 - Operation System Information: Linux 64 bits

#### Installation method
 - from source

**Comment by hobu** [comment]
Is supporting all the way back to gcc 4.9 something that PROJ needs to aspire to? Which platform has this compiler on it? It is very old.

**Comment by SpaceIm** [comment]
I'm packaging proj for conan (in its public repo: https://github.com/conan-io/conan-center-index). gcc4.9 is the oldest compiler for which conan public repo provides binaries.
It's perfectly fine if PROJ 7 decides to not support gcc < 5, I just have to explicitly raise an error in this case.

**Comment by rouault** [comment]
Duplicate of https://github.com/OSGeo/PROJ/issues/2076 already fixed in master and 7.0 maintainance branch
